reframe/core/namespaces.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2016-2021 Swiss National Supercomputing Centre (CSCS/ETH Zurich)
+# ReFrame Project Developers. See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+#
+# Base classes to manage the namespace of a regression test.
+#
+
+
+import abc
+
+
+class LocalNamespace:
+    '''Local namespace of a regression test.
+
+    Temporary storage for test attributes defined in the test class body.
+    This local namespace is populated during the test class body execution.
+
+    Example: In the pseudo-code below, the local namespace of A is {P0},
+    and the local namespace of B is {P1}. However, the final namespace
+    of A is still {P0}, and the final namespace of B is {P0, P1}.
+
+    .. code:: python
+
+        class A(RegressionTest):
+            var('P0')
+
+        class B(A):
+            var('P1')
+    '''
+
+    def __init__(self, namespace=None):
+        self._namespace = namespace or {}
+
+    def __getattr__(self, name):
+        return getattr(self._namespace, name)
+
+    def __getitem__(self, key):
+        return self._namespace[key]
+
+    def __setitem__(self, key, value):
+        if key not in self._namespace:
+            self._namespace[key] = value
+        else:
+            self._raise_namespace_clash(key)
+
+    def __contains__(self, key):
+        return key in self._namespace
+
+    def __iter__(self):
+        return iter(self._namespace)
+
+    def __len__(self):
+        return len(self._namespace)
+
+    def __repr__(self):
+        return f'{type(self).__name__}({self._namespace!r})'
+
+    def _raise_namespace_clash(self, name):
+        '''Raise an error if there is a namespace clash.'''
+        raise ValueError(
+            f'{name!r} is already present in the current namespace'
+        )
+
+
+class Namespace:
```

Comment:
I'm adding the `metaclass=abc.ABCMeta` here. Otherwise the abstract methods below give a very strange error if they're not overridden.